indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveSegmentsToReplaceAction.java

```diff
@@ -0,0 +1,171 @@
+ * the collection only once.
+ *
+ * @implNote This action doesn't produce a {@link Set} because it's implemented via {@link
+ * org.apache.druid.indexing.overlord.IndexerMetadataStorageCoordinator#retrieveUsedSegmentsForIntervals} which returns
+ * a collection. Producing a {@link Set} would require an unnecessary copy of segments collection.
+ */
+public class RetrieveSegmentsToReplaceAction implements TaskAction<Collection<DataSegment>>
+{
+  @JsonIgnore
+  private final String dataSource;
+
+  @JsonIgnore
+  private final Interval interval;
+
+  @JsonCreator
+  public RetrieveSegmentsToReplaceAction(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("interval") Interval interval
+  )
+  {
+    this.dataSource = dataSource;
+    this.interval = interval;
+  }
+
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  public Interval getInterval()
+  {
+    return interval;
+  }
+
+  @Override
+  public TypeReference<Collection<DataSegment>> getReturnTypeReference()
+  {
+    return new TypeReference<Collection<DataSegment>>() {};
+  }
+
+  @Override
+  public Collection<DataSegment> perform(Task task, TaskActionToolbox toolbox)
+  {
+    if (!task.getDataSource().equals(dataSource)) {
+      return toolbox.getIndexerMetadataStorageCoordinator()
+                    .retrieveUsedSegmentsForInterval(dataSource, interval, Segments.ONLY_VISIBLE);
+    }
+
+    final String supervisorId;
+    if (task instanceof AbstractBatchSubtask) {
+      supervisorId = ((AbstractBatchSubtask) task).getSupervisorTaskId();
+    } else {
+      supervisorId = task.getId();
+    }
+
+    final Set<ReplaceTaskLock> replaceLocksForTask = toolbox
+        .getTaskLockbox()
+        .getAllReplaceLocksForDatasource(task.getDataSource())
+        .stream()
+        .filter(lock -> supervisorId.equals(lock.getSupervisorTaskId()))
+        .collect(Collectors.toSet());
+
+    Collection<Pair<DataSegment, String>> segmentsAndCreatedDates =
+        toolbox.getIndexerMetadataStorageCoordinator().retrieveUsedSegmentsAndCreatedDates(dataSource, interval);
+
+    Set<DataSegment> allSegmentsToBeReplaced = new HashSet<>();
+    segmentsAndCreatedDates.forEach(segmentAndCreatedDate -> allSegmentsToBeReplaced.add(segmentAndCreatedDate.lhs));
+    for (Pair<DataSegment, String> segmentAndCreatedDate : segmentsAndCreatedDates) {
+      final DataSegment segment = segmentAndCreatedDate.lhs;
+      final String createdDate = segmentAndCreatedDate.rhs;
+      for (ReplaceTaskLock replaceLock : replaceLocksForTask) {
+        if (replaceLock.getInterval().contains(segment.getInterval())
+            && replaceLock.getVersion().compareTo(createdDate) < 0) {
+          // If a REPLACE lock covers a segment but has a version less than the segment's created date, remove it
+          allSegmentsToBeReplaced.remove(segment);
```

Comment:
Rather than going through the collection of segments and created dates twice (once to add everything, then to remove the invalid ones), start with an empty set and add only the valid entries to it.

Reply:
Done